### PR TITLE
Multi-database search

### PR DESCRIPTION
**Pull request by intsco**
The main purpose of this PR was to add support for search of multiple database formulas combined together to avoid duplicated iso images generation. On the way, I had to completely refactor `msm_basic` subpackage in order to do search faster and lower memory consumption.

* Formula search logic change. Formulas from all databases searched for together, FDR calculated and results stored afterwards by selecting subset of the results for each database separately
* Avoid repeated mz searches and filter images prior to scoring them
* Additional speedup (on big dataset) as a result of replacing image generation and scoring using Spark RDDs with optimized streaming Python+numpy version.
* Search job logic refactored and simplified. As the result computations for all molecular dbs are now done inside MSMSearch.search and the jobs data is stored only after they all successfully finished. If molecule search fails, nothing will be saved to the `job` table (only to `dataset`)
* Centroids generation: set number of centroids per formula to be exactly 4 to get all formula intensities faster by doing a simple array reshape instead of Pandas dataframe grouping
* Formula parser module for parsing and simplifying `formula` +` adduct` as ion formula
* Massive renaming: `sf` -> `formula`, `ion` -> `formula` (or `ion_formula`)

As the formula centroids generation code changed again, all the saved centroids must be deleted before deployment.

The PR is not finished yet as it lacks `S3` support

### CODE REVIEW

**Comment by LachlanStuart** [comment]
>     * Centroids generation: set number of centroids per formula to be exactly 4 to get all formula intensities faster by doing a simple array reshape instead of Pandas dataframe grouping

The added centroids should be filtered out before the annotation is saved to the DB. Otherwise they get saved as a peak at m/z 0:

Old code:
image

This PR:
image

Old code:
image

This PR:
image

**Comment by LachlanStuart** [comment]
Contrary to what we discussed, the decoy adduct random seed is actually a constant `42`. 
Regardless, I compared the same dataset with the same `ds_id` on the old and new code, and these are the annotation counts:

<table>
<tr><th></th><th colspan=2>HMDB-v4</th><th colspan=2>ChEBI-2018-01</th></tr>
<tr><th>FDR</th><th>Before</th><th>After</th><th>Before</th><th>After</th></tr>
<tr><th>&lt;= 5%</th><td>423</td><td>484</td><td>406</td><td>418</td></tr>
<tr><th>&lt;= 10%</th><td>954</td><td>963</td><td>838</td><td>834</td></tr>
<tr><th>&lt;= 20%</th><td>1842</td><td>1835</td><td>1687</td><td>1685</td></tr>
<tr><th>&lt;= 50%</th><td>4083</td><td>4076</td><td>4072</td><td>4064</td></tr>
</table>

Between the 2 DBs:
* 95 annotations' FDRs (95 unique mols) went down by 1 step. None had MSM changes
* 1 annotation's FDR went up by 1 step. None had MSM changes
* 21 annotations (16 unique mols) were only emitted by the old code and not new code. Only 2 of these annotations (/1 unique mol) are explained by the below 3-centroid issue. I'm not sure what has caused the other 19 differences - some of them have quite high MSM scores.
* No annotations moved more than 1 FDR step, except for the annotations that disappeared entirely
* No new annotations were emitted by the new code
* 7 annotations (4 unique mols) had different MSM scores - they all had only 3 centroids, and the msm changed by increasing their spectral score

I have no objection to the FDRs changing. The 3-centroid spectral score issue is probably a bug that needs fixing.

It would be nice to know why some annotations aren't detected in the new code, but I'm out of ideas on how to diagnose it. It doesn't seem to be an issue with segment boundaries - I ran this dataset twice each with the old & new code, and the results were the same, even though all of the runs chose different configurations of segments.

**Comment by intsco** [comment]
The zero mz patterns issue should be resolved in the next commit.

Regarding the decoy adduct random seed. It seems that we discussed the idea of using a fixed one per dataset id but never implemented it because it potentially leads to changes (minor but still) in lots of results when reprocessed.

I'm checking the differences in the search results. The sci test passed successfully.

**Comment by intsco** [comment]
@LachlanStuart I finished cleaning the code. It's still not the final version but I'd like to get your input at this point

File: metaspace/engine/sm/engine/formula_parser.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    def __init__(self, message):
+        self.message = message
+
+
+def parse_formula(f):
+    return [(elem, int(n or '1'))
+            for (elem, n) in formula_regexp.findall(f)]
+
+
+def generate_ion_formula(formula, adduct):
```
[the remark this answers is not in the document]
Comment (intsco) [reply metaspace/engine/sm/engine/formula_parser.py]: Given the fact that this PR is evergrowing I'd propose to add it to one of the following ones

File: metaspace/engine/sm/engine/msm_basic/formula_imager.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        lower_idx = np.searchsorted(sp_mzs, lower, 'l')
+        upper_idx = np.searchsorted(sp_mzs, upper, 'r')
+
+        # Note: consider going in the opposite direction so that
+        # formula_image_metrics can check for the first peak images instead of the last
+        for i, (l, u) in enumerate(zip(lower_idx, upper_idx)):
+            m = None
+            if u - l >= min_px:
+                data = sp_ints[l:u]
+                if data.shape[0] > 0:
```
[the remark this answers is not in the document]
Comment (intsco) [reply metaspace/engine/sm/engine/msm_basic/formula_imager.py]: Good point, I'll fix it

**Pull request merged**